---
<center><h1>Suite Pré-traitement BEDIA JULIEN</h1></center>

---
## Sujet

Dans la continuité du tp précédent, nous allons couvrir les méthodes de prétraitement suivantes:
- Mots vides
- Racinisation/lématisation
- Étiquetage morpho-syntaxique
- Reconnaissance d'entités nommées

Avant tout, un défi est proposé sur les méthodes vues dans le tp précédent.

## Compétences visées

- Découvrir les différentes formes que peuvent revétir un chatbot

### Défi


Dans la section `Lecture de fichiers` du notebook précedent `pre_traitement` et précisément la sous-section `défi`, vous avez chargé des commentaires d'Amazon. Pour chaque commentaire :
- Enlever tous les espaces blancs
- Mettre tous les caractères en minuscules
- Remplacer les URL et les chiffres

Puis trouvez les 60 mots les plus courants.

In [2]:
import os 
import pandas as pd
import re
import nltk
import matplotlib.pyplot as plt
import numpy as np
from collections import Counter

# Expression régulière pour identifier les URLs
url_regex = r'http\S+'

# Expression régulière pour identifier les chiffres
digit_regex = r'\d+'

# Fonction pour faire la subsititution
def replace_pattern_sub(pattern, repl, text):
    return re.sub(pattern, repl, text)

def clean_text(text):
    # Supprimer les espaces blancs
    text = text.strip()

    text = text.lower()
    
    # replace url, chiffre
    text = replace_pattern_sub(url_regex, 'URL', text)
    text = replace_pattern_sub(digit_regex, 'DIGIT', text)


    return text

dossier = "./amazon"
#n_files_to_read = 5
header = []  # initial value for header

contents = []
for i, file_name in enumerate(os.listdir(dossier)):
    file_path = os.path.join(dossier, file_name)
    if not header:
        df = pd.read_csv(file_path, encoding='utf-8')
        header = df.columns.tolist()
    else:
        df = pd.read_csv(file_path, encoding='utf-8', header=None, names=header)
    contents.append(df['Text'])
contents = pd.concat(contents).tolist()

# nettoyage de chaque valeurs dans la liste contents :
contents = [clean_text(text) for text in contents]

# concaténation de tous les textes
all_comments = ''.join(contents)

# tokenisation
PATTERN = r'\w+'
regexpTokenizer = nltk.tokenize.RegexpTokenizer(PATTERN)
tokens = regexpTokenizer.tokenize(all_comments)

# Compter le nombre d'occurrences de chaque mot
word_counts = Counter(tokens)

# Trouver les 5 mots les plus courants
top_words = word_counts.most_common(60)
print("Les 60 mots les plus courants sont : ", top_words)



Les 60 mots les plus courants sont :  [('the', 1870897), ('i', 1681591), ('and', 1289746), ('a', 1235413), ('it', 1072734), ('to', 1015461), ('of', 802044), ('is', 734333), ('this', 670977), ('br', 646925), ('for', 552420), ('in', 547858), ('DIGIT', 519783), ('my', 462143), ('that', 454229), ('but', 376344), ('you', 369377), ('with', 353137), ('have', 345430), ('not', 335004), ('are', 318982), ('was', 315635), ('they', 314438), ('t', 301822), ('s', 301551), ('as', 279663), ('on', 272477), ('like', 256128), ('so', 256057), ('these', 233116), ('them', 211553), ('good', 200223), ('be', 189189), ('or', 181751), ('at', 180137), ('one', 176694), ('can', 174255), ('just', 172825), ('taste', 172652), ('if', 169938), ('very', 167356), ('great', 166583), ('coffee', 166205), ('all', 159729), ('product', 151249), ('flavor', 147836), ('from', 145952), ('we', 144455), ('tea', 138070), ('when', 136243), ('more', 135891), ('has', 131338), ('me', 130561), ('food', 128418), ('had', 127310), ('love', 127

Résultats :

Les 60 mots les plus courants sont :  [('the', 1870897), ('i', 1681591), ('and', 1289746), ('a', 1235413), ('it', 1072734), ('to', 1015461), ('of', 802044), ('is', 734333), ('this', 670977), ('br', 646925), ('for', 552420), ('in', 547858), ('DIGIT', 519783), ('my', 462143), ('that', 454229), ('but', 376344), ('you', 369377), ('with', 353137), ('have', 345430), ('not', 335004), ('are', 318982), ('was', 315635), ('they', 314438), ('t', 301822), ('s', 301551), ('as', 279663), ('on', 272477), ('like', 256128), ('so', 256057), ('these', 233116), ('them', 211553), ('good', 200223), ('be', 189189), ('or', 181751), ('at', 180137), ('one', 176694), ('can', 174255), ('just', 172825), ('taste', 172652), ('if', 169938), ('very', 167356), ('great', 166583), ('coffee', 166205), ('all', 159729), ('product', 151249), ('flavor', 147836), ('from', 145952), ('we', 144455), ('tea', 138070), ('when', 136243), ('more', 135891), ('has', 131338), ('me', 130561), ('food', 128418), ('had', 127310), ('love', 127269), ('will', 127161), ('would', 124295), ('out', 116052), ('than', 114511)]


## Suppression des mots vides

Vous avez peut-être remarqué que les mots les plus courants ci-dessus ne sont pas très intéressants. Il s'agit de mots comme "am", "i", "the" et "a" : des mots vides. Ils nous sont rarement utiles dans l'analyse de texte, et il est donc très courant de les supprimer complètement.

- Quels sont les autres mots vides qui existent dans la langue anglaise?

Pour répondre à cette question, afficher la liste des mots vides du module `nltk.corpus` (voir documentation officielle de `nltk`). 

In [3]:
import nltk
nltk.download('stopwords')

from nltk.corpus import stopwords

stop_words = stopwords.words('english')

print(stop_words)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Julien\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


### Défi

Utilisez la liste de ces mots vides anglais pour nettoyer de mots vides les commentaires Amazon chargés au-dessus.

In [4]:
filtered_words = [word for word in tokens if word.lower() not in stop_words]

# Compter le nombre d'occurrences de chaque mot restant
word_counts = Counter(filtered_words)

# Trouver les 60 mots les plus courants
top_words = word_counts.most_common(60)

print(top_words)

[('br', 646925), ('DIGIT', 519783), ('like', 256128), ('good', 200223), ('one', 176694), ('taste', 172652), ('great', 166583), ('coffee', 166205), ('product', 151249), ('flavor', 147836), ('tea', 138070), ('food', 128418), ('love', 127269), ('would', 124295), ('get', 108740), ('really', 101012), ('much', 93093), ('amazon', 89898), ('also', 86860), ('time', 84609), ('use', 83880), ('little', 83492), ('buy', 76808), ('best', 76665), ('tried', 76455), ('price', 75786), ('even', 74959), ('find', 73513), ('well', 73235), ('make', 71997), ('better', 70653), ('dog', 69817), ('try', 69155), ('eat', 67717), ('cup', 62856), ('first', 61937), ('water', 61585), ('chocolate', 61079), ('bag', 58195), ('sugar', 57324), ('used', 54403), ('found', 54116), ('sweet', 52647), ('made', 51315), ('drink', 51144), ('box', 50786), ('free', 50658), ('bought', 50241), ('way', 49547), ('two', 49455), ('think', 49162), ('day', 48792), ('since', 46511), ('go', 46322), ('store', 45878), ('tastes', 45579), ('still', 

Les 60 mots les plus courants :

[('br', 646925), ('DIGIT', 519783), ('like', 256128), ('good', 200223), ('one', 176694), ('taste', 172652), ('great', 166583), ('coffee', 166205), ('product', 151249), ('flavor', 147836), ('tea', 138070), ('food', 128418), ('love', 127269), ('would', 124295), ('get', 108740), ('really', 101012), ('much', 93093), ('amazon', 89898), ('also', 86860), ('time', 84609), ('use', 83880), ('little', 83492), ('buy', 76808), ('best', 76665), ('tried', 76455), ('price', 75786), ('even', 74959), ('find', 73513), ('well', 73235), ('make', 71997), ('better', 70653), ('dog', 69817), ('try', 69155), ('eat', 67717), ('cup', 62856), ('first', 61937), ('water', 61585), ('chocolate', 61079), ('bag', 58195), ('sugar', 57324), ('used', 54403), ('found', 54116), ('sweet', 52647), ('made', 51315), ('drink', 51144), ('box', 50786), ('free', 50658), ('bought', 50241), ('way', 49547), ('two', 49455), ('think', 49162), ('day', 48792), ('since', 46511), ('go', 46322), ('store', 45878), ('tastes', 45579), ('still', 45441), ('could', 44975), ('order', 44595), ('know', 43616)]

##  Etiquettage morpho-syntaxique

L'étiquettage morpho-syntaxique consiste à assigner à chaque token les informations grammaticales associées (par exemple un nom, un verbe, un adjectif, etc.). NLTK propose plusieurs étiquetteurs. Nous nous focalisons sur l'étiquetteur `pos_tag`. Lors de l'étiquetage, il est conseillé de ne pas supprimer les mots vides au préalable. 

Choisissez les 10 premiers commentaires Amazon que vous avez déjà chargés puis appliquez l'opération de tokénisation. Ensuite utiliser l'étiquetteur `pos_tag` sur chacun de ces commentaires.

In [5]:
from nltk.tag import pos_tag

n_10_comments = (contents[:10])

# nettoyage de chaque valeurs dans la liste contents :
n_10_comments_clean = [clean_text(text) for text in n_10_comments]

# concaténation de tous les textes
n_10_comments = ''.join(n_10_comments_clean)

# tokenisation
PATTERN = r'\w+'
regexpTokenizer = nltk.tokenize.RegexpTokenizer(PATTERN)
tokens = regexpTokenizer.tokenize(n_10_comments)

nltk.download('averaged_perceptron_tagger')

pos_tag(tokens)

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\Julien\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


[('i', 'NNS'),
 ('have', 'VBP'),
 ('bought', 'VBN'),
 ('several', 'JJ'),
 ('of', 'IN'),
 ('the', 'DT'),
 ('vitality', 'NN'),
 ('canned', 'VBD'),
 ('dog', 'JJ'),
 ('food', 'NN'),
 ('products', 'NNS'),
 ('and', 'CC'),
 ('have', 'VBP'),
 ('found', 'VBN'),
 ('them', 'PRP'),
 ('all', 'DT'),
 ('to', 'TO'),
 ('be', 'VB'),
 ('of', 'IN'),
 ('good', 'JJ'),
 ('quality', 'NN'),
 ('the', 'DT'),
 ('product', 'NN'),
 ('looks', 'VBZ'),
 ('more', 'RBR'),
 ('like', 'IN'),
 ('a', 'DT'),
 ('stew', 'NN'),
 ('than', 'IN'),
 ('a', 'DT'),
 ('processed', 'JJ'),
 ('meat', 'NN'),
 ('and', 'CC'),
 ('it', 'PRP'),
 ('smells', 'VBZ'),
 ('better', 'RBR'),
 ('my', 'PRP$'),
 ('labrador', 'NN'),
 ('is', 'VBZ'),
 ('finicky', 'JJ'),
 ('and', 'CC'),
 ('she', 'PRP'),
 ('appreciates', 'VBZ'),
 ('this', 'DT'),
 ('product', 'NN'),
 ('better', 'JJR'),
 ('than', 'IN'),
 ('most', 'JJS'),
 ('product', 'NN'),
 ('arrived', 'VBD'),
 ('labeled', 'VBN'),
 ('as', 'IN'),
 ('jumbo', 'NN'),
 ('salted', 'VBD'),
 ('peanuts', 'NNS'),
 ('the',

### Défi 

Charger le contenu du fichier 'sowing-and-reaping.txt'. Prétraitez-le en suivant les étapes suivantes :
    
- Supprimez les espaces blancs
- Remplacez tous les chiffres par '0'
- Convertissez le résultat en tokens
- Marquez chaque token avec une étiquette morpho-synthaxique

In [6]:

# Ouvrir le fichier
with open('sowing-and-reaping.txt', 'r', encoding='utf-8') as file:
    text = file.read()

# Clean text
text.strip()
text = text.lower()
text = replace_pattern_sub(digit_regex, '0', text)

# Tokenize
tokens = regexpTokenizer.tokenize(text)

# POS tagging
pos_tag(tokens)

[('the', 'DT'),
 ('project', 'NN'),
 ('gutenberg', 'NN'),
 ('ebook', 'NN'),
 ('sowing', 'NN'),
 ('and', 'CC'),
 ('reaping', 'NN'),
 ('by', 'IN'),
 ('frances', 'NNS'),
 ('ellen', 'VBN'),
 ('watkins', 'NNS'),
 ('harper', 'RBR'),
 ('edited', 'VBN'),
 ('by', 'IN'),
 ('frances', 'NNS'),
 ('smith', 'RB'),
 ('foster', 'RBR'),
 ('this', 'DT'),
 ('ebook', 'NN'),
 ('is', 'VBZ'),
 ('for', 'IN'),
 ('the', 'DT'),
 ('use', 'NN'),
 ('of', 'IN'),
 ('anyone', 'NN'),
 ('anywhere', 'RB'),
 ('at', 'IN'),
 ('no', 'DT'),
 ('cost', 'NN'),
 ('and', 'CC'),
 ('with', 'IN'),
 ('almost', 'RB'),
 ('no', 'DT'),
 ('restrictions', 'NNS'),
 ('whatsoever', 'IN'),
 ('you', 'PRP'),
 ('may', 'MD'),
 ('copy', 'VB'),
 ('it', 'PRP'),
 ('give', 'VB'),
 ('it', 'PRP'),
 ('away', 'RB'),
 ('or', 'CC'),
 ('re', 'VB'),
 ('use', 'IN'),
 ('it', 'PRP'),
 ('under', 'IN'),
 ('the', 'DT'),
 ('terms', 'NNS'),
 ('of', 'IN'),
 ('the', 'DT'),
 ('project', 'NN'),
 ('gutenberg', 'NN'),
 ('license', 'NN'),
 ('included', 'VBD'),
 ('with', 'IN'),

## Racinisation/Lemmatisation

La racinisation et la lemmatisation consistent toutes deux à supprimer les affixes morphologiques des mots. 

La racinisation consiste à tronquer un mot de toute déclinaison, inflexions et dérivations pour n'obtenir que la racine. En traitement automatique des langues, on opère la racinisation, souvent, en supprimant une partie de la fin du mot. La taille de la fin du mot dépend de la méthode de racinisation utilisée. La taille peut être sous-estimée ou sur-estimée. En conséquence la racine pourrait ne pas exister dans le dictionnaire de la langue concernée.

La lemmatisation consiste à réduire un mot à sa forme de base, quels que soient ses déclinaisons et ses accords. La forme de base dépend de la langue. En français, on choisit l'infinitif ou le genre, mais pas systématiquement. 

Nous procédons à la racinisation et la lemmatisation parce que, souvent, nous nous soucions davantage du contenu essentiel du mot. Par exemple, on ne soucie pas d'un verbe à la troisième personne du présent, ou du futur mais plutôt de son sens. Néanmoins, si la tâche TAL nécessite de connaître le temps du verbe, la lemmatisation doit être personnalisée. La même remarque s'applique à la racinisation.

NLTK fournit de nombreux algorithmes pour la racinisation et la lemmatisation. Des bases de références sont les algorithmes de racinisation `PorterStemmer` et `SnowballStemmer` et l'algorithme de lemmatisation `WordNetLemmatizer`.

Soit les mots suivants walks, eating, corrected, appliquer l'une des deux algorithmes de racinisation. Les racines obtenues sont-elles valides linguistiquement (autrement dit, ont elles unt existence réelle dans la langue anglaise?).

In [7]:
words = ['walks', 'eating', 'corrected']

# Racinisation avec PorterStemmer
porter = nltk.stem.PorterStemmer()
porter_stems = [porter.stem(word) for word in words]
print(porter_stems)

['walk', 'eat', 'correct']


Les mots ont un sens dans la langue anglaise après racinisation : ['walk', 'eat', 'correct']

Soit les mots suivants mice, wolves, continued, women, theses, crises, criteria, appliquer les deux méthodes de racinisations. Qu'observez-vous?

In [8]:
words = ['mice', 'wolves', 'continued', 'women', 'theses', 'crises', 'criteria']

# Racinisation avec PorterStemmer
porter_stems = [porter.stem(word) for word in words]
print("Racinisation avec PorterStemmer", porter_stems)

# Racinisation avec SnowballStemmer
snowball = nltk.stem.SnowballStemmer('english')
snowball_stems = [snowball.stem(word) for word in words]
print("Racinisation avec SnowballStemmer", snowball_stems)

Racinisation avec PorterStemmer ['mice', 'wolv', 'continu', 'women', 'these', 'crise', 'criteria']
Racinisation avec SnowballStemmer ['mice', 'wolv', 'continu', 'women', 'these', 'crise', 'criteria']


Les deux méthodes de racinisations sont sensiblements identiques.
Cependant, nous perdons le sens de certains mots lorsqu'on l'applique sur des noms, notamment 'wolves' qui devient 'wolv'
La racinisation récupère seulement la racine d'un mot.

Les mots ne sont pas mis au singulier ou ne sont pas des mots de la langue anglaise.

Appliquer un lemmatiseur sur cette même liste. Qu'observez-vous?

In [9]:
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')

words = ['mice', 'wolves', 'continued', 'women', 'theses', 'crises', 'criteria']

# Lemmatizer avec WordNet
lemmatizer = nltk.stem.WordNetLemmatizer()
lemmatized_words = [lemmatizer.lemmatize(word) for word in words]
print("Lemmatizer avec WordNet", lemmatized_words)

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Julien\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\Julien\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


Lemmatizer avec WordNet ['mouse', 'wolf', 'continued', 'woman', 'thesis', 'crisis', 'criterion']


Les mots ont été transformés par leur singulier sauf 'continued' qui est resté le même.

Que s'est-il passé avec le terme `continued`? Il est parfois nécessaire de fournir au lemmatiseur une étiquette morpho-syntaxique pour l'aider à résoudre les cas ambigus. WordNetLemmatizer a un paramètre `pos` permettant de préciser la classe du mot. Parmi les valeurs possibles de ce paramètre sont:
- "n" pour les noms,
- "v" pour les verbes,
- "a" pour les adjectifs, 
- "r" pour les adverbes

Utilisez ce paramètre pour résoudre le cas de `continued`.

In [10]:
word = 'continued'

# Lemmatizer avec WordNet
lemmatizer = nltk.stem.WordNetLemmatizer()
lemmatized_continued = lemmatizer.lemmatize(word, pos='v')
print("Lemmatizer de 'continued' : ", lemmatized_continued)

Lemmatizer de 'continued' :  continue


Utiliser ce paramètre pour le terme `leaves` en tant que verbe et nom. Qu'observez-vous?

In [11]:
word = 'leaves'

# Lemmatizer avec WordNet
lemmatizer = nltk.stem.WordNetLemmatizer()
lematized_v = lemmatizer.lemmatize(word, pos='v')
print("Lemmatizer de 'leaves' en tant que verbe : ", lematized_v)
lematized_n = lemmatizer.lemmatize(word, pos='n')
print("Lemmatizer de 'leaves' en tant que nom : ", lematized_n)

Lemmatizer de 'leaves' en tant que verbe :  leave
Lemmatizer de 'leaves' en tant que nom :  leaf


Le mot 'leaves' correspond à la troisième personne du singulier du verbe 'leave' et au pluriel du mot 'leaf'. Il peut être interpreté de deux façon différentes.

Spacy est une bibliothèque dédiée au traitement automatique des langues. Elle propose un lemmatiseur basé sur des modèles d'apprentissage automatique pré-entraînés capables de déterminer étiquettes morpho-syntaxiques et produire des lemmes par un simple appel de fonction, contrairement au lemmatiseur WordNetLemmatize, où les étiquettes morpho-syntaxiques doivent être explicitement fournies.

Nous pouvons installer Spacy et télécharger le modèle `en` pour la langue anglaise en exécutant la commande suivante en ligne de commande: 

pip install spacy && python -m spacy download en

Charger le modèle `en` que vous affecterez à la variable `nlp`. Cette variable est, en réalité, une fonction.


In [12]:
import spacy

# Charger le modèle 'en' pour la langue anglaise
nlp = spacy.load('en_core_web_sm')

Appelez cette fonction sur la phrase suivante : She said she was a "fighter not a quitter" as she tried to reassert authority over the fraught ranks of her Conservative Party on Wednesday, amid growing media reports that momentum was building to oust her.



In [13]:
text = "She said she was a 'fighter not a quitter' as she tried to reassert authority over the fraught ranks of her Conservative Party on Wednesday, amid growing media reports that momentum was building to oust her."

nlp_output = nlp(text)

# Afficher le texte et le POS tag pour chaque token
for token in nlp_output:
    print(token.text, token.pos_)


She PRON
said VERB
she PRON
was AUX
a DET
' PUNCT
fighter NOUN
not PART
a DET
quitter NOUN
' PUNCT
as SCONJ
she PRON
tried VERB
to PART
reassert VERB
authority NOUN
over ADP
the DET
fraught ADJ
ranks NOUN
of ADP
her PRON
Conservative PROPN
Party PROPN
on ADP
Wednesday PROPN
, PUNCT
amid ADP
growing VERB
media NOUN
reports NOUN
that SCONJ
momentum NOUN
was AUX
building VERB
to PART
oust VERB
her PRON
. PUNCT


### Défi

Utilisez les algorithmes de racinisation et les algorithmes de lemmatisation sur chacun des tweets (voir notebook précédent). Vous devez au préalable avoir segmenté les tweets en phrases et supprimé les mots vides. A noter que vous devez utiliser, avec le lemmatiseur WordNetLemmatize, un étiquetteur morpho-syntaxique sur une liste de tokens issue d'une phrase. 

Identifiez les mots uniques des résultats de chacun des quatre algorithmes puis calculer le nombre de termes communs pour chaque paire des résultats. 

In [39]:
import pandas as pd

def unique_words(tableau):
    unique = set()
    for ligne in tableau:
        for mot in ligne:
            unique.add(mot)
    return list(unique)

tweets = pd.read_csv('airline_tweets.csv', encoding='utf-8')

# On transforme les tweets en token
tweets_token = tweets['text'][:10000].apply(nltk.word_tokenize)

tweets_clean = []
for tweets in tweets_token:
    words_filter = [word for word in tweets if word.lower() not in stop_words]
    tweets_clean.append(words_filter)


# On utilise les algos de lemmatisation et de racinisation (SnowBall, Porter, Spacy et WordNet)
wordNet = list()
snowball = list()
porter = list()
spacy = []
for tweet in tweets_clean:
    wordNet.append([nltk.WordNetLemmatizer().lemmatize(word) for word in tweet])
    snowball.append([nltk.SnowballStemmer('english').stem(word) for word in tweet])
    porter.append([nltk.PorterStemmer().stem(word) for word in tweet])
    spacy.append(nlp(' '.join(tweet)))

# On affiche les résultats
# print("WordNet : ", wordNet)
# print("Snowball : ", snowball)
# print("Porter : ", porter)
# print("Spacy : ", spacy)

# Création d'un ensemble de mots uniques pour chaque tableau de phrases
spacyTweets = set(unique_words(spacy))
wordNetTweets = set(unique_words(wordNet))
snowBallTweets = set(unique_words(snowball))
porterTweets = set(unique_words(porter))


# Calcul de l'intersection des ensembles de mots entre spacy et les autres
mots_communs_spacy_wordnet = spacyTweets.intersection(wordNetTweets)
mots_communs_spacy_snowball = spacyTweets.intersection(snowBallTweets)
mots_communs_spacy_porter = spacyTweets.intersection(porterTweets)

# Affichage du nombre de mots communs
print("Sur " + str(len(tweets_token)) + " tweets, voici la comparaison des mots communs entre : ")
print("Spacy - WordNet : ", len(mots_communs_spacy_wordnet))
print("Spacy - SnowBall : ", len(mots_communs_spacy_snowball))
print("Spacy - Porter : ", len(mots_communs_spacy_porter))





Sur 10000 tweets, voici la comparaison des mots communs entre : 
Spacy - WordNet :  0
Spacy - SnowBall :  0
Spacy - Porter :  0


## Reconnaissance d'entités nommées(REN)

Lorsque nous lisons et traitons des phrases, nous avons tendance à identifier d'abord les acteurs clés de la phrase (par exemple, les personnes, les lieux et les organisations). Cette classification nous aide à décomposer la phrase en entités et à donner un sens à la sémantique de la phrase. La reconnaissance d'entités nommées imite le même comportement et est utilisée pour classer les entités nommées (ou noms propres) dans un texte donné. Les applications de cette catégorisation sont largement utilisées dans l'industrie. L'une des applications réelles de REN est le résumé de texte. Celui-ci consiste à analyser des documents textuels et à résumer ceux-ci en identifiant les entités clés du document. Un cas d'utilisation populaire est la catégorisation de CV, dans lequel le REN traite un grand nombre de CV et met en évidence les entités clés telles que le nom, l'institution et les compétences, ce qui facilite une évaluation rapide. 

Appelez la fonction `nlp` sur la phrase précédente "She said she was...". Vous assignez le résultat dans une variable appelée `res`. Cette dernière a un attribut ents correspondant à une liste d'objets. Dans l'objet, vous récupérer l'entité nommée avec l'attribut label_ et le mot ou composé de mots associé avec l'attribut text. 

Créez une liste de tuples où chacune contient les attributs test et label_.



In [40]:
resultat = nlp('She said she was a "fighter not a quitter" as she tried to reassert authority over the fraught ranks of her Conservative Party on Wednesday, amid growing media reports that momentum was building to oust her.')


tuples = []

# On récupère dans resultat les entités label_ et text
for entite in resultat.ents:
    tuples.append((entite.label_, entite.text))

# On affiche les résultats
print(tuples)

[('ORG', 'Conservative Party'), ('DATE', 'Wednesday')]


## Défi

Reprenez les commentaires Amazon pre-traités et appliquez REN. Puis organisez les entités nommées par type (par exemple, organisation, personne...). Les types sont fournis dans les attributs label_. 

Calculez pour chaque type, les 10 premiers textes les plus fréquents. 



In [53]:
import os
import pandas as pd

NB_FICHIER = 5
NB_COMMENTAIRES_TRAITES = 100

# Récupération des fichiers amazon
dossier = "./amazon"
header = []  # initial value for header

contents = []
for i, file_name in enumerate(os.listdir(dossier)[:NB_FICHIER]):
    file_path = os.path.join(dossier, file_name)
    if not header:
        df = pd.read_csv(file_path, encoding='utf-8')
        header = df.columns.tolist()
    else:
        df = pd.read_csv(file_path, encoding='utf-8', header=None, names=header)
    contents.append(df['Text'])
contents = pd.concat(contents).tolist()

# Nettoyage de chaque valeurs dans la liste contents :
contents = [clean_text(text) for text in contents]

########################################

# On applique REN pour récupérer les entités nommées et les organiser par type graçe à l'attribut label_
res = [nlp(text) for text in contents[:NB_COMMENTAIRES_TRAITES]]
entities_by_type = {}
for ents in res:
    for ent in ents.ents:
        if ent.label_ not in entities_by_type:
            entities_by_type[ent.label_] = []
        entities_by_type[ent.label_].append(ent.text)

# Pour chaque type d'entité, on calcule les 10 entités les plus fréquentes, de la même manière qu'avec les mots question 1
for entity_type in entities_by_type:
    entity_counts = Counter(entities_by_type[entity_type])
    most_frequent_entities = entity_counts.most_common(10)
    print(f"Type d'entité : {entity_type}")
    print(most_frequent_entities)
    print("------------------------------")

Type d'entité : DATE
[('christmas', 2), ('halloween', 2), ('around a few centuries', 1), ('only two weeks', 1), ('more than two years', 1), ('july', 1), ('all these years', 1), ('several years', 1), ('/>two years ago', 1), ('decades', 1)]
------------------------------
Type d'entité : GPE
[('vermont', 2), ('c.s', 1), ('pennsylvania', 1), ('the united states', 1), ('mexico', 1), ('america', 1), ('newton', 1), ('texas', 1), ('nirvana', 1)]
------------------------------
Type d'entité : QUANTITY
[('six pounds', 2), ('five pound', 1), ('a half cup', 1), ('several hundred mile', 1)]
------------------------------
Type d'entité : ORDINAL
[('second', 4), ('first', 3), ('firstly', 1), ('third', 1)]
------------------------------
Type d'entité : CARDINAL
[('one', 7), ('two', 3), ('six', 2), ('three', 2), ('almost half', 1), ('five', 1), ('zero', 1), ('half', 1), ('though.<br', 1)]
------------------------------
Type d'entité : ORG
[('mccann', 8), ('maple & brown sugar', 3), ('y & s candies, inc